cmd/dep/godep_config.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/golang/dep"
+	"github.com/golang/dep/internal/gps"
+	"github.com/pkg/errors"
+)
+
+type godepFile struct {
+	json    godepJson
+	loggers *dep.Loggers
+}
+
+func newGodepFile(loggers *dep.Loggers) *godepFile {
+	return &godepFile{loggers: loggers}
+}
+
+type godepJson struct {
+	Name    string         `json:"ImportPath"`
+	Imports []godepPackage `json:"Deps"`
+}
+
+type godepPackage struct {
+	ImportPath string `json:"ImportPath"`
+	Rev        string `json:"Rev"`
+	Comment    string `json:"Comment"`
+}
+
+// load parses Godeps.json in projectDir and unmarshals the json to godepFile.json
+func (g *godepFile) load(projectDir string) error {
+	j := filepath.Join(projectDir, "Godeps", godepJsonName)
+	if g.loggers.Verbose {
+		g.loggers.Err.Printf("godep: Loading %s", j)
+	}
+
+	raw, err := ioutil.ReadFile(j)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to read %s", j)
+	}
+	err = json.Unmarshal(raw, &g.json)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to parse %s", j)
+	}
+
+	return nil
+}
+
+func (g *godepFile) convert(projectName string, sm gps.SourceManager) (*dep.Manifest, *dep.Lock, error) {
+	// Create empty manifest and lock
+	manifest := &dep.Manifest{
+		Dependencies: make(gps.ProjectConstraints),
+	}
+	lock := &dep.Lock{}
+
+	// Parse through each import and add them to manifest and lock
+	for _, pkg := range g.json.Imports {
+		// ImportPath must not be empty
+		if pkg.ImportPath == "" {
+			err := errors.New("godep: Invalid godep configuration, ImportPath is required")
+			return nil, nil, err
+		}
+
+		// Rev must not be empty
+		if pkg.Rev == "" {
+			err := errors.New("godep: Invalid godep configuration, Rev is required")
+			return nil, nil, err
+		}
+
+		if pkg.Comment == "" {
+			// When there's no comment, try to get corresponding version for the Rev
+			// and fill Comment.
+			// Get all the versions
+			pi := gps.ProjectIdentifier{ProjectRoot: gps.ProjectRoot(pkg.ImportPath)}
+			versions, err := sm.ListVersions(pi)
+			if err != nil {
+				return nil, nil, err
+			}
+			// Sort the versions in descending order, newer versions first
+			gps.SortPairedForUpgrade(versions)
+			// Match Rev with versions' underlying revision
+			for _, v := range versions {
+				if string(v.Underlying()) == pkg.Rev {
+					pkg.Comment = v.String()
+					break
+				}
+			}
+		}
```

Comment:
There is - have a look at `version_unifier.go`. Actually, there's a whole pile of logic for this, but it's all hidden away unexported inside gps, as it's encoded as an extension to the `Version` system, and I didn't want to leak an extra type out there.

I'd strongly prefer we not leak that unifier type publicly - note the unconditional panics in certain of the interface's method implementations - but we could expose some alternate frontend to it.

I've also been thinking that we might want to export the `source` interface (and from there, `sourceGateway`s), and include such methods on it. That's probably out of scope for immediate needs, though...